engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java

```diff
@@ -135,13 +146,103 @@ private Either<List<DirectBuffer>, CatchEventTuple> findCatchEventInScope(
+      final var instanceRecord = instance.getValue();
+      final var process = getProcess(instanceRecord.getProcessDefinitionKey());
+
+      final var found = findEscalationCatchEventInProcess(escalationCode, process, instance);
+      if (found.isPresent()) {
+        return found;
+      }
+
+      // find in parent process instance if exists
+      final var parentElementInstanceKey = instanceRecord.getParentElementInstanceKey();
+      instance = elementInstanceState.getInstance(parentElementInstanceKey);
+    }
+
+    // no matching catch event found
+    return Optional.empty();
+  }
+
+  private Optional<CatchEventTuple> findEscalationCatchEventInProcess(
+      final DirectBuffer escalationCode,
+      final ExecutableProcess process,
+      ElementInstance instance) {
+
+    while (instance != null && instance.isActive() && !instance.isInterrupted()) {
+      final var found = findEscalationCatchEventInScope(escalationCode, process, instance);
+      if (found.isPresent()) {
+        return found;
+      }
+      // find in parent scope if exists
+      final var instanceParentKey = instance.getParentKey();
+      instance = elementInstanceState.getInstance(instanceParentKey);
+    }
 
     // no matching catch event found
     return Optional.empty();
   }
 
+  private Optional<CatchEventTuple> findEscalationCatchEventInScope(
+      final DirectBuffer escalationCode,
+      final ExecutableProcess process,
+      final ElementInstance instance) {
+    final var processInstanceRecord = instance.getValue();
+    final var elementId = processInstanceRecord.getElementIdBuffer();
+    final var elementType = processInstanceRecord.getBpmnElementType();
+
+    final var element = process.getElementById(elementId, elementType, ExecutableActivity.class);
+    if (element.getEvents().isEmpty()) {
+      return Optional.empty();
+    }
+
+    final List<ExecutableCatchEvent> catchEvents =
+        element.getEvents().stream().filter(ExecutableCatchEvent::isEscalation).toList();
+
+    final List<ExecutableCatchEvent> catchStartEvents =
+        catchEvents.stream().filter(ExecutableStartEvent.class::isInstance).toList();
+    final List<ExecutableCatchEvent> catchBoundaryEvents =
+        catchEvents.stream().filter(ExecutableBoundaryEvent.class::isInstance).toList();
+
+    final boolean foundCatchStartEvent =
+        findEscalationCatchEvent(
+            instance,
+            catchStartEvents,
+            catchEvent -> findCatchEventPredicate.test(catchEvent, escalationCode));
+
+    if (foundCatchStartEvent) {
+      return Optional.of(catchEventTuple);
+    }
+
+    final boolean foundCatchBoundaryEvent =
+        findEscalationCatchEvent(
+            instance,
+            catchBoundaryEvents,
+            catchEvent -> findCatchEventPredicate.test(catchEvent, escalationCode));
+
+    if (foundCatchBoundaryEvent) {
+      return Optional.of(catchEventTuple);
+    }
+
+    return Optional.empty();
```

Comment:
❌ We can replace this all with a simple stream. We don't have to care if an event is a start event or a boundary event. As long as we are searching for the events in the correct order, meaning from the scope closest to the throw event to the scope farthest away from the throw event, we can return the first event we find regardless of its event type.

```java
    final Optional<ExecutableCatchEvent> catchEvent =
        element.getEvents().stream()
            .filter(ExecutableCatchEvent::isEscalation)
            .filter(
                event -> {
                  final var eventEscalationCode = event.getEscalation().getEscalationCode();
                  return eventEscalationCode.capacity() == 0
                      || eventEscalationCode.equals(escalationCode);
                })
            .findFirst();
```

This should be enough to find the correct catch event. For readability's sake we could pull the 2nd filter to a separate method to make it more descriptive. Something along the lines of `matchesEscalationCode(event, escalationCode)`.

If the optional contains a catchEvent we can set the tuple and return an optional of it. Otherwise we must return an empty optional.